### ISSUE TITLE
errors during run the codes provided in the paper

### ISSUE THREAD

**Issue by byronwwang** [torchnet/torchnet#7]
File.lua:141: Unwritable object <function> at <?>.callback.closure.mnist.testdataset.createdataset.readlush.torch.cat

**Comment by byronwwang** [comment]
When I check the torchnet codes, I found that, in listdataset.lua,  [line 116](https://github.com/torchnet/torchnet/blob/master/dataset/listdataset.lua#L116), for different of list, it should be different. When list is a table, #self.list should be a number, but when list is LongTensor , #self.list is a tensor. Am I right? Maybe this caused the problem.

**Comment by lvdmaaten** [comment]
No this does look like a serialization issue that arises when you're creating the `ParallelDatasetIterator`. This creates a number of threads; at thread construction time, upvalues are serialized. Can you post a repro? Also, please make sure your `threads` package is up-to-date.

**Comment by FCDT** [comment]
Hello,
I have the same issue when running on fresh Torch+Torchnet install:

When running the code as is:

File.lua:141: Unwritable object <function> at <?>.callback.closure.mnist.testdataset.createdataset.readlush.torch.cat

If I try to define the datasets outside of the getIterator (to circumvent the syntax trick used that is not "transferrable" to other datasets
example:
local trainset = mnist['traindataset']()
local testset = mnist.testdataset()

and then assigning one or the other to the local dataset inside getIterator
Example:
local function loadmnist(mode)
    local mnist = require 'mnist'
    local dataset = {}
    if mode == 'train' then
        dataset = mnist.traindataset()
    else
    dataset = mnist.testdataset()
    end
    return dataset
end

then 
local function getIterator(mode)
    return tnt.ParallelDatasetIterator {
        nthread = 1,
        init = function() require 'torchnet' end,
        closure = function()
            local dataset = loadmnist(mode)
            return tnt.BatchDataset {
                batchsize = 128,
                dataset = tnt.ListDataset {
                    list = torch.range(1, dataset.data:size(1)),
                    load = function(idx)
                        return {
                            input = dataset.data[idx],
                            target = torch.LongTensor {
                                dataset.label[idx]
                            },
                        } -- sample contains input and target
                    end,
                }
            }
        end,
    }
end
I get:
___
invalid arguments!
stack traceback:
	[C]: in function 'error'
	[string "argcheck"]:355: in function '__init'
	/home/x/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/x/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'ListDataset'
	test_torchnet.lua:18: in function 'closure'
	...are/lua/5.1/torchnet/dataset/paralleldatasetiterator.lua:71: in function <...are/lua/5.1/torchnet/dataset/paralleldatasetiterator.lua:70>
	[C]: in function 'xpcall'
	...x/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
	...e/x/torch/install/share/lua/5.1/threads/queue.lua:65: in function <...e/x/torch/install/share/lua/5.1/threads/queue.lua:41>
	[C]: in function 'pcall'
	...e/x/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
	[string "  local Queue = require 'threads.queue'..."]:13: in main chunk
stack traceback:
	[C]: in function 'error'
	...x/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
	...x/torch/install/share/lua/5.1/threads/threads.lua:264: in function 'synchronize'
	...x/torch/install/share/lua/5.1/threads/threads.lua:142: in function 'specific'
	...x/torch/install/share/lua/5.1/threads/threads.lua:125: in function 'Threads'
	...are/lua/5.1/torchnet/dataset/paralleldatasetiterator.lua:77: in function '__init'
	/home/x/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/x/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'getIterator'
	test_torchnet.lua:54: in main chunk
	[C]: at 0x00406670


Calling list = torch.totable(torch.range(1, dataset.data:size(1))) solves the problem but then gives

/home/x/torch/install/bin/luajit: /home/x/torch/install/share/lua/5.1/nn/Container.lua:67: 
/home/x/torch/install/share/lua/5.1/nn/Linear.lua:69: input must be vector or matrix
stack traceback:


as output
In 1 module of nn.Sequential:

**Comment by shreyshahi** [comment]
Hi, I am facing the same problem with `ParallelDatasetIterator`. I did a clean new install of `torch` and installed `threads`. Can any other library cause the serialization problem?

**Comment by lvdmaaten** [comment]
Apparently, the `mnist` package is not serializable, perhaps because it contains objects that live in C-land, which is why you cannot do this. The solution is to not serialize the `mnist` object, by putting the content of `loadmnist` in the `closure` function (this way, the `mnist` package will be loaded inside each thread instead of being serialized).

**Comment by gooshan** [comment]
As hinted by @byronwwang there is an actual bug in listdataset.lua, however, it is on line [126](https://github.com/torchnet/torchnet/blob/master/dataset/listdataset.lua#L126). When a torch.LongTensor is provided the assert statement fails with "

> both (null) and torch.LongTensor have no less-equal operator"

. 
The assert statement should be modified for the case of LongTensor by using `:size(1)` instead of `#`. The second option, which I am using, is to convert my list from `LongTensor` to Lua `table` with `torch.totable()`

**Comment by lvdmaaten** [comment]
@gooshan I think I fixed that bug a few minutes ago.

**Comment by lvdmaaten** [comment]
It appears the example in the paper was done for a different version of the `mnist` package. This code example should work fine: 

```
local tnt   = require 'torchnet'

local function getIterator(mode)
  return tnt.ParallelDatasetIterator{
    nthread = 1,
    init    = function() require 'torchnet' end,
    closure = function()
      local mnist = require 'mnist'
      local dataset = mnist[mode .. 'dataset']()
      dataset.data = dataset.data:reshape(
         dataset.data:size(1),
         dataset.data:size(2) *
         dataset.data:size(3)
      ):double()
      return tnt.BatchDataset{
         batchsize = 128,
         dataset = tnt.ListDataset{
           list = torch.range(
             1, dataset.data:size(1)
           ):long(),
           load = function(idx)
             return {
               input  = dataset.data[idx],
               target = torch.LongTensor{
                 dataset.label[idx] + 1
               },
             } -- sample contains input and target
           end,
        }
      }
    end,
  }
end

local net = nn.Sequential():add(nn.Linear(784,10))

local engine = tnt.SGDEngine()
local meter  = tnt.AverageValueMeter()
local clerr  = tnt.ClassErrorMeter{topk = {1}}
engine.hooks.onStartEpoch = function(state)
  meter:reset()
  clerr:reset()
end
engine.hooks.onForwardCriterion =
function(state)
  meter:add(state.criterion.output)
  clerr:add(
    state.network.output, state.sample.target)
  print(string.format(
    'avg. loss: %2.4f; avg. error: %2.4f',
    meter:value(), clerr:value{k = 1}))
end

local criterion = nn.CrossEntropyCriterion()

engine:train{
  network   = net,
  iterator  = getIterator('train'),
  criterion = criterion,
  lr        = 0.1,
  maxepoch  = 10,
}
engine:test{
  network   = net,
  iterator  = getIterator('test'),
  criterion = criterion,
}
```

I will add it to the codebase.

**Comment by lvdmaaten** [comment]
See https://github.com/torchnet/torchnet/blob/master/example/mnist.lua for a working example with comments.
